addons/hr_expense/models/hr_expense_sheet.py

```diff
@@ -464,6 +464,23 @@ def create(self, vals_list):
         sheets.activity_update()
         return sheets
 
+    def write(self, values):
+        res = super().write(values)
+        # Ensures there is no empty expense report in a state different from draft or cancel
+        if 'state' in values and values['state'] or 'expense_line_ids' in values or 'approval_state' in values:
+            for sheet in self:
+                state = sheet.state
+                if sheet.expense_line_ids:
+                    continue
+                if state in {'submit', 'approve', 'post', 'done'} or sheet.approval_state:
+                    if 'expense_line_ids' in values and not sheet.expense_line_ids:
+                        raise UserError(_("You cannot remove all expenses from a submitted, approved or paid expense report."))
+                    else:
+                        raise UserError(_(
+                            "This expense report is empty. You cannot submit or approve an empty expense report."
+                        ))
+        return res
```

Comment:
Could you explain to me exactly what are the conditions we are checking and why.
I'm a bit confused here.
Maybe let's chat directly in vocal. That will be easier and faster.

Reply:
If the (approval)? state is modified -> Check you have expense lines
If you remove all expense lines -> Check the state allows it